### PR TITLE
Add DialURL to allow connection using a Redis URL

### PR DESCRIPTION
**Pull request by shosti**
After thinking about it some more, I think it's cleaner to take a URL
object instead of an unparsed string, but I think either way would work.

### CODE REVIEW

File: redis/conn.go [hunk 1]
```diff
@@ -104,6 +116,14 @@ func DialDatabase(db int) DialOption {
 	}}
 }
 
+// DialPassword specifies the password to use when connecting to
+// redis.
```
Comment (garyburd) [inline redis/conn.go]: "when connecting to the Redis server."

File: redis/conn.go [hunk 2]
```diff
@@ -143,6 +163,55 @@ func Dial(network, address string, options ...DialOption) (Conn, error) {
...
+// scheme (https://www.iana.org/assignments/uri-schemes/prov/redis).
+func DialURL(u *url.URL, options ...DialOption) (Conn, error) {
+	if u.Scheme != "redis" {
+		return nil, fmt.Errorf("invalid redis URL scheme: %s", u.Scheme)
+	}
+
+	// As per the IANA draft spec, the host defaults to localhost and
+	// the port defaults to 6379
+	host, port, err := net.SplitHostPort(u.Host)
+	if err != nil {
```
Comment (garyburd) [inline redis/conn.go]: Parsing error text is fragile. Assume that the port is not specified if err != nil.  

    if err != nil {
        host = u.Host
        port = "6379"
    } else if host == "" {
        host = "localhost"
    }

File: redis/conn_test.go [hunk 3]
```diff
@@ -422,6 +424,105 @@ func TestReadDeadline(t *testing.T) {
...
+func TestDialURL(t *testing.T) {
+	var (
+		u   *url.URL
+		err error
+	)
+	for _, d := range dialErrors {
+		u, _ = url.Parse(d.rawurl)
+		_, err = redis.DialURL(u)
+		if err == nil || !strings.Contains(err.Error(), d.expectedError) {
+			t.Errorf("NewDialer did not throw correct error (expected %v to contain %s)", err, d.expectedError)
```
Comment (garyburd) [inline redis/conn_test.go]: Replace "throw" with "return".  Errors are not thrown in Go.

**Comment by garyburd** [comment]
Use a string to specify the URL.  It's simpler for applications.  This is not a scenario where *url.URL should be used to avoid the design flaw in the url.URL type.

To avoid destroying user data when running tests, check to see if the database is empty before writing to the database.  Is there a way to check the selected database without writing to the database?

See inline comments on code.

File: redis/conn.go (hunk 2 above)
Comment (shosti) [inline redis/conn.go]: I did originally, but there are some actual parse errors that are useful to return (for instance, if there are too many colons). I'll change it back, though, I guess invalid hosts will eventually be caught trying to connect.

**Comment by shosti** [comment]
OK, I updated the pull request. Unfortunately, I don't think there's any way to check which database is selected without writing to the database (correct me if I'm wrong), so I just added a check to make sure that DBSIZE is 0. Let me know if you notice anything else.

**Pull request merged**